File: crates/contracts-ws/src/envelope.rs
//! WS wire envelope — Rust mirror of `contracts/ws/envelope.go`.
//!
//! The Go file is canonical (shipped cycle 21 L4.L); we serialize JSON
//! that is byte-for-byte compatible.

use serde::{Deserialize, Serialize};

/// Current wire format version. Bumping is a cross-language change.
pub const ENVELOPE_VERSION: u8 = 1;

/// Classifies an envelope as control vs data.
///
/// Routers fast-path control messages (`ws.ping` / `ws.pong` /
/// `ws.refresh`) without hitting the authz cache (S12 §12AB.4).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum MessageKind {
    /// Protocol/control messages. Bypasses S2/S3 authz.
    Control,
    /// Application messages — subject to S2/S3 authz on every send/receive.
    Data,
}

/// Direction marker — server-side validators reject mismatches.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Direction {
    /// Inbound to the gateway.
    #[serde(rename = "c2s")]
    ClientToServer,
    /// Outbound to the browser.
    #[serde(rename = "s2c")]
    ServerToClient,
}

/// The wire-shape every WS frame deserializes to. Wire-compatible with
/// `contracts/ws/envelope.go::Envelope` (Go canonical).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Envelope {
    /// Envelope schema version. Must equal [`ENVELOPE_VERSION`].
    #[serde(rename = "v")]
    pub version: u8,

    /// Control vs data.
    pub kind: MessageKind,

    /// Message-type string (e.g., "chat.message", "ws.ping").
    #[serde(rename = "type")]
    pub message_type: String,

    /// Direction marker.
    #[serde(rename = "dir")]
    pub direction: Direction,

    /// Monotonic per-connection per-type sequence.
    #[serde(default, skip_serializing_if = "is_zero_u64")]
    pub seq: u64,

    /// UUID string; server tracks in TTL set for replay defense.
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub nonce: String,

    /// Opaque per-type bytes.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub payload: Option<serde_json::Value>,
}

#[inline]
fn is_zero_u64(v: &u64) -> bool {
    *v == 0
}

impl Envelope {
    /// Construct a server→client control envelope.
    pub fn control(message_type: impl Into<String>) -> Self {
        Self {
            version: ENVELOPE_VERSION,
            kind: MessageKind::Control,
            message_type: message_type.into(),
            direction: Direction::ServerToClient,
            seq: 0,
            nonce: String::new(),
            payload: None,
        }
    }

    /// Construct a server→client data envelope.
    pub fn data(message_type: impl Into<String>, nonce: impl Into<String>, payload: serde_json::Value) -> Self {
        Self {
            version: ENVELOPE_VERSION,
            kind: MessageKind::Data,
            message_type: message_type.into(),
            direction: Direction::ServerToClient,
            seq: 0,
            nonce: nonce.into(),
            payload: Some(payload),
        }
    }

    /// Shape validator — mirrors `contracts/ws/envelope.go::Validate`.
    pub fn validate(&self) -> Result<(), EnvelopeError> {
        if self.version != ENVELOPE_VERSION {
            return Err(EnvelopeError::VersionMismatch {
                got: self.version,
                want: ENVELOPE_VERSION,
            });
        }
        if self.message_type.is_empty() {
            return Err(EnvelopeError::EmptyType);
        }
        if matches!(self.kind, MessageKind::Data) && self.nonce.is_empty() {
            return Err(EnvelopeError::NonceRequired);
        }
        Ok(())
    }
}

/// Envelope-level validation errors.
#[derive(Debug, thiserror::Error)]
pub enum EnvelopeError {
    /// Wire version does not match this build.
    #[error("envelope version mismatch: got {got}, want {want}")]
    VersionMismatch {
        /// Version received on the wire.
        got: u8,
        /// Version this build expects.
        want: u8,
    },
    /// Type field is empty.
    #[error("envelope type empty")]
    EmptyType,
    /// Data envelope without nonce.
    #[error("data envelope requires nonce (replay defense)")]
    NonceRequired,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn version_pinned_at_1() {
        assert_eq!(ENVELOPE_VERSION, 1);
    }

    #[test]
    fn json_round_trip_matches_go_canonical_shape() {
        // This is the wire shape `contracts/ws/envelope.go` serializes.
        // If either side drifts, the JSON below will diverge — pinning
        // the field names guards against accidental rename.
        let env = Envelope::data("chat.message", "n1", serde_json::json!({"body": "hi"}));
        let s = serde_json::to_string(&env).unwrap();
        assert!(s.contains("\"v\":1"));
        assert!(s.contains("\"kind\":\"data\""));
        assert!(s.contains("\"type\":\"chat.message\""));
        assert!(s.contains("\"dir\":\"s2c\""));
        assert!(s.contains("\"nonce\":\"n1\""));

        let back: Envelope = serde_json::from_str(&s).unwrap();
        assert_eq!(back.message_type, "chat.message");
        assert_eq!(back.direction, Direction::ServerToClient);
        assert_eq!(back.kind, MessageKind::Data);
        assert!(back.validate().is_ok());
    }

    #[test]
    fn validate_rejects_version_mismatch() {
        let mut env = Envelope::control("ws.ping");
        env.version = 99;
        assert!(matches!(env.validate(), Err(EnvelopeError::VersionMismatch { .. })));
    }

    #[test]
    fn validate_rejects_data_without_nonce() {
        let env = Envelope {
            version: ENVELOPE_VERSION,
            kind: MessageKind::Data,
            message_type: "chat.message".into(),
            direction: Direction::ServerToClient,
            seq: 1,
            nonce: String::new(),
            payload: Some(serde_json::json!({"body": "hi"})),
        };
        assert!(matches!(env.validate(), Err(EnvelopeError::NonceRequired)));
    }
}

File: contracts/ws/envelope.go
package ws

import (
	"encoding/json"
	"errors"
	"fmt"
)

// EnvelopeVersion is the current wire format version. Bumping is a
// cross-language change (Go + Rust mirror in lockstep per Q-L4-1).
const EnvelopeVersion = 1

// MessageKind classifies an envelope as control vs data. Routers
// fast-path control messages (ws.ping / ws.pong / ws.refresh) without
// hitting the authz cache (S12 §12AB.4).
type MessageKind string

const (
	// KindControl — protocol/control messages: ws.ping, ws.pong,
	// ws.refresh, ws.close. Bypasses S2/S3 authz.
	KindControl MessageKind = "control"

	// KindData — application messages: chat.*, session.*, presence.*,
	// event.delivery. Subject to S2/S3 authz on every send/receive.
	KindData MessageKind = "data"
)

// Direction marks who sent the envelope. Used by server-side validators
// to reject server→client message types the client tries to send (and
// vice versa).
type Direction string

const (
	// DirectionClientToServer — inbound to the gateway (chat.message,
	// session.join, ws.refresh, etc.).
	DirectionClientToServer Direction = "c2s"

	// DirectionServerToClient — outbound to the browser (event.delivery,
	// ws.close, presence.update, etc.).
	DirectionServerToClient Direction = "s2c"
)

// Envelope is the wire-shape every WS frame deserializes to. Both
// inbound (client→server) and outbound (server→client) frames use
// the same envelope; Direction + Type let the router dispatch.
//
// The envelope is INTENTIONALLY minimal — payload is opaque JSON
// (gateway routes by Type; downstream service code parses Payload
// per its known schema). Bumping EnvelopeVersion is a wire-contract
// change that requires Rust + browser-lib parity.
type Envelope struct {
	// Version pins the envelope schema. Must equal EnvelopeVersion at
	// runtime; mismatch is a hard reject (close code 4010).
	Version int `json:"v"`

	// Kind = control or data (see MessageKind).
	Kind MessageKind `json:"kind"`

	// Type — message-type string (e.g., "chat.message", "ws.ping",
	// "session.kick"). Foundation does NOT constrain the namespace
	// beyond non-empty; L6 ws-gateway owns the canonical V1 list per
	// S12 §12AB.8.
	Type string `json:"type"`

	// Direction — c2s or s2c (server-side validator rejects mismatches).
	Direction Direction `json:"dir"`

	// Seq — monotonic per-connection per-Type sequence. Server rejects
	// duplicates + out-of-order (tolerance per S12 §12AB.7). Zero is
	// reserved for control messages where seq doesn't apply.
	Seq uint64 `json:"seq,omitempty"`

	// Nonce — UUID string; server tracks in TTL set (60s) for replay
	// defense. Required for data messages; optional for control.
	Nonce string `json:"nonce,omitempty"`

	// Payload — opaque per-type bytes. The gateway routes by Type and
	// hands Payload to the consuming service intact.
	Payload json.RawMessage `json:"payload,omitempty"`
}

// Validate enforces envelope-level shape invariants. Per-Type field
// validation is owned by the L6 router (cycle 27+).
func (e Envelope) Validate() error {
	if e.Version != EnvelopeVersion {
		return fmt.Errorf("ws: envelope version %d != current %d", e.Version, EnvelopeVersion)
	}
	switch e.Kind {
	case KindControl, KindData:
	default:
		return fmt.Errorf("ws: invalid envelope kind %q", e.Kind)
	}
	switch e.Direction {
	case DirectionClientToServer, DirectionServerToClient:
	default:
		return fmt.Errorf("ws: invalid envelope direction %q", e.Direction)
	}
	if e.Type == "" {
		return errors.New("ws: envelope type empty")
	}
	if e.Kind == KindData && e.Nonce == "" {
		return errors.New("ws: data envelope requires nonce (replay defense)")
	}
	if e.Kind == KindData && e.Seq < 1 {
		// ws/v1.yaml: seq is `minimum: 1`, required for data, omitted for
		// control. Seq 0 on a data frame is reserved/invalid (it would also
		// trip WSSession.AcceptSeq's "seq=0 reserved for control" guard).
		return errors.New("ws: data envelope requires seq>=1")
	}
	return nil
}

// CloseCode is one of the enumerated S12 §12AB.9 close codes. Wire
// values match WebSocket spec: 1000 + 4001..4010.
type CloseCode uint16

const (
	// CloseNormal — 1000 client-initiated normal closure.
	CloseNormal CloseCode = 1000

	// CloseTokenExpired — 4001 refresh failed before expiry.
	CloseTokenExpired CloseCode = 4001

	// CloseTokenRevoked — 4002 user logout / JWT revoked.
	CloseTokenRevoked CloseCode = 4002

	// CloseUserErased — 4003 S8 crypto-shred fired.
	CloseUserErased CloseCode = 4003

	// CloseRealityArchived — 4004 S10 reality state archived/dropped.
	CloseRealityArchived CloseCode = 4004

	// CloseAdminKick — 4005 S5 Tier 2 Griefing action.
	CloseAdminKick CloseCode = 4005

	// CloseRateLimitExceeded — 4006 persistent L5 violation.
	CloseRateLimitExceeded CloseCode = 4006

	// CloseOriginMismatch — 4007 L4 violation mid-connection.
	CloseOriginMismatch CloseCode = 4007

	// CloseConnectionLimitExceeded — 4008 L5 per-user LRU eviction.
	CloseConnectionLimitExceeded CloseCode = 4008

	// CloseFingerprintMismatch — 4009 L6 client binding broken.
	CloseFingerprintMismatch CloseCode = 4009

	// CloseSchemaInvalid — 4010 persistent malformed messages.
	CloseSchemaInvalid CloseCode = 4010
)

// AllCloseCodes returns every enumerated close code in numeric order.
// Used by tests + the future cross-language conformance lint to
// confirm Go + Rust + (eventually) browser TS lib stay in sync.
func AllCloseCodes() []CloseCode {
	return []CloseCode{
		CloseNormal,
		CloseTokenExpired,
		CloseTokenRevoked,
		CloseUserErased,
		CloseRealityArchived,
		CloseAdminKick,
		CloseRateLimitExceeded,
		CloseOriginMismatch,
		CloseConnectionLimitExceeded,
		CloseFingerprintMismatch,
		CloseSchemaInvalid,
	}
}

// IsValid returns true iff c is one of the enumerated close codes.
// The gateway MUST refuse to close with any other code (defense
// against typos / silent V2 drift).
func (c CloseCode) IsValid() bool {
	for _, ok := range AllCloseCodes() {
		if c == ok {
			return true
		}
	}
	return false
}

// String returns the canonical short name (matches §12AB.9 vocab).
// Used by structured logs + admin tooling.
func (c CloseCode) String() string {
	switch c {
	case CloseNormal:
		return "normal_closure"
	case CloseTokenExpired:
		return "token_expired"
	case CloseTokenRevoked:
		return "token_revoked"
	case CloseUserErased:
		return "user_erased"
	case CloseRealityArchived:
		return "reality_archived"
	case CloseAdminKick:
		return "admin_kick"
	case CloseRateLimitExceeded:
		return "rate_limit_exceeded"
	case CloseOriginMismatch:
		return "origin_mismatch"
	case CloseConnectionLimitExceeded:
		return "connection_limit_exceeded"
	case CloseFingerprintMismatch:
		return "fingerprint_mismatch"
	case CloseSchemaInvalid:
		return "schema_invalid"
	}
	return fmt.Sprintf("invalid_close_code(%d)", uint16(c))
}
